### ISSUE TITLE
Guidelines for evaluating defense using Auto Attack

### ISSUE THREAD

**Issue by SohamTamba** [fra31/auto-attack#43]
Hello, what is the protocol for using Auto Attack to evaluate a Defense?
For example, it has been found [here](https://arxiv.org/pdf/1802.00420.pdf) that Shattered Gradients and Stochastic Gradients could result in a false sense of security.
The [README](https://github.com/fra31/auto-attack/blob/master/README.md) mentions one should use `version='rand'` to combat Stochastic Gradients, but what is the protocol for Shattered Gradients? I know that Square Attack will not be affected by Shattered Gradients since it uses Black Box access only.

For example, what if I train a ResNet50 to solve CIFAR-10, and then attach a `GradientKiller` layer at the bottom? How will Auto Attack compute adversarial examples then?

Code in Pytorch:
```
class GradientKiller(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input):
        return input

    @staticmethod
    def backward(ctx, grad_output):
        return torch.zeros(grad_output.size())

class SuperGeniusDefense(torch.nn.Module):
    def __init__(self, trained_model):
        super(SuperGeniusDefense, self).__init__()
        self.model = trained_model

    def forward(X):
        X = GradientKiller.apply(X)
        X = self.model(X)
        return X
```

Now providing the `AutoAttack` adversary with the forward pass of `SuperGeniusDefense` will provide a false sense of security when it is actually not safer than `SuperGeniusDefense.model`.

**Comment by fra31** [comment]
Hi,

as you mentioned using Expectation over Transformation (EoT) is the main tool to evaluate randomized defenses.
In case a defense doesn't allow the gradient to be computed correctly, it depends on which is the source of missing gradients. In general, I think black-box attacks are a good starting point: if evaluating a new defense Square Attack is more effective than the white-box attacks, then it is likely that the gradient is not informative and might make sense to increase number of iterations and restarts given to Square.

However, in the specific case you propose, the simple solution is to directly attack the `SuperGeniusDefense.model` and then use the found adversarial points on `SuperGeniusDefense`. Clearly, if one applies blindly any white-box attack relying on the gradient to a model which returns zero gradients everywhere, those will fail. Since we evaluate robustness in the white-box scenario, the internal parameters and the output of intermediate layers are available to the attackers, and in this case should be exploited.
Currently, deciding when using the output of the penultimate layer instead of the last one cannot be done automatically by AutoAttack, but it is up to the user to provide the correct function.

It is also important to note that the case you mention doesn't help to improve robustness, but just masks the vulnerability of the classifiers. For example, https://arxiv.org/abs/2010.09670 discusses some criteria a model should satisfy in order to prevent phenomena as those you describe.

Hope this clarifies things!

**Comment by SohamTamba** [comment]
Thanks, that helps.
I was wondering if there is any API I can use to automatically adapt to Shattered Gradients, so I don't have to worry that the attack I used to evaluate my proposed defense is too weak (it is not always as obvious as `GradientKiller`), but I don't think that is available.
